src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -442,37 +466,72 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
             logger.atDebug().log("Fail to add publish request to spooler queue", e);
             future.completeExceptionally(e);
             return future;
         }
-        return CompletableFuture.completedFuture(0);
+        return spooledMessage.getFuture();
+    }
+
+    protected boolean isValidPublishRequest(PublishRequest request) throws SpoolerStoreException {
+        // Payload size should be smaller than MQTT maximum message size
+        int messageSize = request.getPayload().length;
+        if (messageSize > MAXIMUM_SIZE_OF_MQTT_MESSAGE_SIZE) {
+            throw new SpoolerStoreException(String.format("The message size exceeds the maximum size of %d ",
+                    MAXIMUM_SIZE_OF_MQTT_MESSAGE_SIZE));
+        }
+
+        // Topic should not contain wildcard characters
+        String topic = request.getTopic();
+        if (topic.contains("#") || topic.contains("+")) {
+            throw new SpoolerStoreException("The topic of publish request should not contain wildcard characters.");
+        }
+
+        // Topic should not have no more than maximum number of forward slashes (/)
+        if (topic.chars().filter(num -> num == '/').count() > MAXIMUM_NUMBER_OF_FORWARD_SLASHES) {
+            throw new SpoolerStoreException(String.format("The topic of publish request should not contain "
+                    + "more than %d forward slashes.", MAXIMUM_NUMBER_OF_FORWARD_SLASHES));
+        }
+        return true;
     }
 
     @SuppressWarnings({"PMD.AvoidCatchingThrowable", "PMD.PreserveStackTrace"})
     protected CompletableFuture<Integer> publishSingleSpoolerMessage() throws InterruptedException {
         long id = -1L;
         try {
             id = spool.popId();
-            PublishRequest request = spool.getMessageById(id);
+            SpoolMessage spooledMessage = spool.getMessageById(id);
+            PublishRequest request = spooledMessage.getRequest();
             MqttMessage m = new MqttMessage(request.getTopic(), request.getPayload());
 
             long finalId = id;
             return getConnection(false).publish(m, request.getQos(), request.isRetain())
                     .whenComplete((packetId, throwable) -> {
                         // packetId is the SDK assigned ID. Ignore this and instead use the spooler ID
                         if (throwable == null) {
+                            spooledMessage.getFuture().complete(0);
                             spool.removeMessageById(finalId);
                             logger.atDebug().kv("id", finalId).kv("topic", request.getTopic())
                                     .log("Successfully published message");
                         } else {
-                            spool.addId(finalId);
-                            logger.atError().log("Failed to publish the message via Spooler and will retry",
-                                    throwable);
+                            if (!mqttOnline.get() || spooledMessage.getRetried().get()
```

Comment:
the online check seems to not be useful since IoT Core drops us for most of the bad packet types. And in cases where sending a message doesn't cause us to drop, then we don't want to hold that against the message.